### PR TITLE
Prevent non hyper-v admin users to execute machine commands

### LINKED ISSUE #27614: Prevent non hyper-v admin users to execute machine commands
### Issue Description

We used to prevent users to run machine commands from non privileged terminal, but after #26277 that's not enforced anymore. That made sense because that's not required anymore, but, to start, stop and list hyper-v VMs a user need to belong to the Hyper-V administrator group and that should be checked now  

### Steps to reproduce the issue

Steps to reproduce the issue
1. Set hyper-v as the provider: `$env:CONTAINERS_MACHINE_PROVIDER="hyperv"`
2. Run a machine command: `podman machine start/stop/list/info/inspect/cp/ssh/set`

### Describe the results you received

The machine commands return an cryptic error:

```
Error: no results found
```

### Describe the results you expected

Instead, the command should fail with something like:

```
You need to belong to the hyper-v administrators to run this command
```

### PR DESCRIPTION
**Pull request by lstocchi**
Following the work done in #26277 (which allowed init and remove operations for the "Hyper-V Administrators" group), this PR extends that logic to all actions targeting Hyper-v machines.

it resolves #27614 

#### Checklist

Ensure you have completed the following checklist for your pull request to be reviewed:

- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [ ] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

```release-note
Permission validation for Hyper-V has been extended to cover all actions. Users in the "Hyper-V Administrators" group can now execute all actions without needing mandatory full local Administrator privileges. Users without the right privileges will now see an error message when attempting to execute these commands.
```

### CODE REVIEW

**Comment by mheon** [comment]
@l0rd PTAL

File: pkg/machine/provider/platform_test.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ func TestSupportedProviders(t *testing.T) {
 	case "darwin":
 		assert.Equal(t, []define.VMType{define.AppleHvVirt, define.LibKrun}, SupportedProviders())
 	case "windows":
-		assert.Equal(t, []define.VMType{define.WSLVirt, define.HyperVVirt}, SupportedProviders())
+		assert.Equal(t, []define.VMType{define.HyperVVirt, define.WSLVirt}, SupportedProviders())
```
Comment (mheon) [inline pkg/machine/provider/platform_test.go]: Why the ordering swap?

File: pkg/machine/provider/platform_windows.go [hunk 2]
```diff
@@ -15,6 +15,12 @@ import (
 	"go.podman.io/common/pkg/config"
 )
 
+// Variables to hold permission check functions for testing purposes.
+var (
+	hasAdminRightsFunc       = windows.HasAdminRights
+	hasHyperVAdminRightsFunc = hyperv.HasHyperVAdminRights
```
Comment (mheon) [inline pkg/machine/provider/platform_windows.go]: These don't seem to add much as written. It seems like a single functioning combining both would be more useful, and line 86 could just be `windows.HasAdminRights()`

File: pkg/machine/provider/platform_windows.go [hunk 3]
```diff
@@ -39,17 +45,26 @@ func GetByVMType(resolvedVMType define.VMType) (vmconfigs.VMProvider, error) {
...
 	return nil, fmt.Errorf("unsupported virtualization provider: `%s`", resolvedVMType.String())
 }
 
 func GetAll() []vmconfigs.VMProvider {
-	return []vmconfigs.VMProvider{
+	providers := []vmconfigs.VMProvider{
 		new(wsl.WSLStubber),
-		new(hyperv.HyperVStubber),
 	}
+	if hasAdminRightsFunc() || hasHyperVAdminRightsFunc() {
```
Comment (baude) [inline pkg/machine/provider/platform_windows.go]: I am struggling with the logic here if I am a user.  If I don't have permissions, should it not exist?  Earlier, when we returned an error, that certainly made sense to me.

File: pkg/machine/provider/platform_test.go (hunk 1 above)
Comment (lstocchi) [reply pkg/machine/provider/platform_test.go]: Bc `SupportedProviders()` returns `[]define.VMType{define.HyperVVirt, define.WSLVirt}` and `assert.Equal` performs a strict comparison so the items order matters. It failed the way it was written.

File: pkg/machine/provider/platform_windows.go (hunk 3 above)
Comment (lstocchi) [reply pkg/machine/provider/platform_windows.go]: They have two different purposes imo. Above (`GetByVMType()`) returns an error because user is explicitly executing a command and if they do not have rights to do so we stop them and return an error.
`GetAll` is more a discovery thing. It is used by the `list` command. If someone wants to list machines, it's annoying to show an error message, maybe they're just interested to WSL. So if you don't have permissions to work with Hyper-V it skips it. Hyper-v console works the same - if you don't have the rights you see nothing.

Comment (l0rd) [reply pkg/machine/provider/platform_windows.go]: This is about `podman machine ls`, and the problem has changed in 6.0. If the user doesn't have privileges we still want to list WSL machines.

But we should indeed fail when the user specifies `--provider hyperv`. And if the default provider is `hyperv` too.

File: pkg/machine/provider/platform_test.go (hunk 1 above)
Comment (l0rd) [reply pkg/machine/provider/platform_test.go]: And CI isn't catching it because `pkg\machine\provider` is [skipped](https://github.com/containers/podman/blob/3681055601c5c2a882fc6773a2fc72ef6eb1dadf/winmake.ps1#L81C23-L81C43).  @lstocchi would you mind removing that package from the skipped ones?

File: pkg/machine/provider/platform_windows.go (hunk 2 above)
Comment (l0rd) [reply pkg/machine/provider/platform_windows.go]: @lstocchi you could add a new `hyperv.CanList()` function similar to the `hyperv.canCreate()` where you hide the logic.

File: pkg/machine/provider/platform_test.go (hunk 1 above)
Comment (mheon) [reply pkg/machine/provider/platform_test.go]: Can't you use `assert.ElementsMatch` or similar? I dislike having tests be this fragile.

Comment (lstocchi) [reply pkg/machine/provider/platform_test.go]: @mheon @l0rd done both changes

File: pkg/machine/provider/platform_windows.go (hunk 2 above)
Comment (lstocchi) [reply pkg/machine/provider/platform_windows.go]: done. I named it `HasHyperVPermissions` as it's more generic and not just related to `list`, let me know

**Comment by baude** [comment]
@lstocchi will you be addressing the problems @l0rd mentions here or @l0rd do we need to plan for this?

else LGTM

**Comment by l0rd** [comment]
From my point of view this PR is now mergeable as is. It's up to @lstocchi, if he has bandwidth, to fix those corner cases. Otherwise I will open an issue and we can prioritize them later. Let's wait for him to come back (today is national holiday but tomorrow he should be back).

**Comment by lstocchi** [comment]
> I did some tests and all work as expected with the expection of some corner cases.
> 
> Inspecting or starting a WSL machine, if the default provider is hyperv, fails:
> 
> ```powershell
> > $env:CONTAINERS_MACHINE_PROVIDER="hyperv"
> > podman.exe machine inspect mywslmachine`
> Error: Hyper-V machines require Hyper-V admin rights...
> > podman.exe machine start mywslmachine`
> Error: Hyper-V machines require Hyper-V admin rights...
> ```
> 
> And this is awkward because `podman machine ls` works and shows the WSL machines. But that may not be related to this PR.

I think this is a change that would affect all OSes with more than 1 provider, not just Windows. I'd rather have a separate PR to test it also on macOS.

> Also commands like `machine cp`, `machine ssh` etc..., against an hyperv machine, should work too, even if the user doesn't have Hyper-V admin privileges.
> 
> That said these scenarios aren't blockers from my point of view.

I'm going to add the change here as it seems related to this PR

**Comment by l0rd** [comment]
After discussion we decided that it was ok to merge the PR as is. I have created a separate issue https://github.com/containers/podman/issues/27723

/approve /lgtm
